pandas/core/groupby/ops.py

```diff
@@ -914,7 +914,8 @@ def agg_series(
         np.ndarray or ExtensionArray
         """
 
-        if not isinstance(obj._values, np.ndarray):
+        # if objtype is not in np.dtypes, type is preserved
+        if not isinstance(obj._values, np.ndarray) and obj.dtype != "boolean":
```

Comment:
I don't think this is the right fix. While the issue raised is with `boolean`, it can occur with any dtype.

One possible solution is to take `npvalues` and convert it to the corresponding `dtype_backend` - e.g. if it starts out as a `numpy_nullable`, then we can do:

```
from pandas.core.dtypes.cast import convert_dtypes
            
dtype = convert_dtypes(npvalues, dtype_backend="numpy_nullable")
out = Series(npvalues, dtype=dtype)
```

Similarly for the `pyarrow` backend. This seems to me to be the right behavior because it parallel's the inference we're doing with NumPy. But I can't say I feel very confident there aren't some cases where this would still go wrong.

cc @jbrockmendel @WillAyd for any thoughts.

Reply:
Not sure what the best alternative would be but that's unfortunate we'd have to convert pyarrow boolean types to numpy

Reply:
still reading, but first thought is that instead of comparing to the string "boolean" the check should be `isinstance(dtype, BooleanDtype)`

Reply:
OK, looking at the test makes it clear what the underlying issue is, and its a tough one.  maybe_cast_pointwise_result goes through _from_scalars, which it turns out was a mis-feature (xref #56430).  What we need is an EA method that does family-preserving inference (i.e. if you start with pyarrow/masked/numpy/sparse, you end up with pyarrow/masked/numpy/sparse).

Shorter-term, it looks like BooleanArray._from_scalars is insufficiently strict; it should raise when it sees floats.

Reply:
`maybe_convert_objects` already has a `convert_to_nullable_dtype` - but only for bool/int. Could this be expanded upon - both for other nullable dtypes as well as pyarrow? It's already quite large and complex, but it seems to me there is an advantage of having a single function to call when you are in the situation of "I have an object NumPy array, and need to determine how to do type inference".

That being said, if the EA method @jbrockmendel is proposing completely supplants `maybe_convert_objects` in the long term, that sounds like a good route to me.

Reply:
> but im wary of introducing even more ways

No disagreement there. It just seems like your proposed EA method from https://github.com/pandas-dev/pandas/pull/58258#discussion_r1567668218 needs to go through all the values, determine what is seen, come up with the output dtype, and then make it so. The first two steps are done in `maybe_convert_objects`, and I'd be wary of duplicating that logic elsewhere.

Reply:
i suspect that what you're describing is likely to be the basis of the MaskedArray._whatever implementation.